modules/host/contractmanager/writeaheadlogsync.go

```diff
@@ -0,0 +1,281 @@
+	threadsStopped := make(chan struct{})
+	syncLoopStopped := make(chan struct{})
+	wal.syncChan = make(chan struct{})
+	go wal.threadedSyncLoop(threadsStopped, syncLoopStopped)
+	wal.cm.tg.AfterStop(func() {
+		// Close the threadsStopped channel to let the sync loop know that all
+		// calls to tg.Add() in the contract manager have cleaned up.
+		close(threadsStopped)
+
+		// Because this is being called in an 'AfterStop' routine, all open
+		// calls to the contract manager should have completed, and all open
+		// threads should have closed. The last call to change the contract
+		// manager should have completed, so the number of uncommitted changes
+		// should be zero.
+		<-syncLoopStopped // Wait for the sync loop to signal proper termination.
+
+		// Perform a final sync to make sure all open ended changes get to
+		// disk. This call is necessary because open-ended resources are not
+		// sync'd during a commit - instead their sync is delayed until the
+		// next call to commit() to maximize disk throughput and minimzie lock
+		// contention.
+		wal.syncResources()
+
+		// Close the dangling commit resources.
+		err = wal.fileWALTmp.Close()
+		if err != nil {
+			wal.cm.log.Println("Error closing wal file during contract manager shutdown:", err)
+		}
+		err = os.Remove(filepath.Join(wal.cm.persistDir, walFileTmp))
+		if err != nil {
+			wal.cm.log.Println("Error removing temporary WAL during contract manager shutdown:", err)
+		}
+		// Allow the removal of the WAL file to be disrupted to enable easier
+		// simluations of power failure.
+		if !wal.cm.dependencies.disrupt("cleanWALFile") {
+			err = os.Remove(filepath.Join(wal.cm.persistDir, walFile))
+			if err != nil {
+				wal.cm.log.Println("Error removing WAL during contract manager shutdown:", err)
+			}
+		}
+		for _, sf := range wal.cm.storageFolders {
+			err = sf.file.Close()
+			if err != nil {
+				wal.cm.log.Println("Error closing the storage folder file handle", err)
+			}
+		}
+	})
+	return nil
+}
+
+// threadedSyncLoop is a background thread that occasionally commits the WAL to
+// the state as an ACID transaction. This process can be very slow, so
+// transactions to the contract manager are batched automatically and
+// occasionally committed together.
+func (wal *writeAheadLog) threadedSyncLoop(threadsStopped chan struct{}, syncLoopStopped chan struct{}) {
+	// Provide a place for the testing to disable the sync loop.
+	if wal.cm.dependencies.disrupt("threadedSyncLoopStart") {
+		close(syncLoopStopped)
+		return
+	}
+
+	syncInterval := 1500 * time.Millisecond
+	for {
+		select {
+		case <-threadsStopped:
+			close(syncLoopStopped)
+			return
+		case <-time.After(syncInterval):
+			// Commit all of the changes in the WAL to disk, and then apply the
+			// changes. Measure how long it takes to apply the changes, and use
+			// that to steer the amount of time that should be waited to the
+			// next sync.
+			wal.mu.Lock()
+			wal.commit()
+			wal.mu.Unlock()
```

Comment:
The lock is to protect `wal.commit()` lol, which does a ton of state modifications.